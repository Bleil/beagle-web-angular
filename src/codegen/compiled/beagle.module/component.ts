/* 
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/

import { Type } from '@angular/core'
import { kebabToCamelCase, removeExtraIndentation } from '../../utils/formatting'
import { getComponentAnnotations } from '../../utils/metadata'
import { remoteViewSelector } from '../../../constants'


function createViewChildString(name: string, templateName: string, angularVersion: number): string {
  if (angularVersion >= 8) {
    return `'${name}': new ViewChild('${templateName}', { static: true })`
  } else {
    return `'${name}': new ViewChild('${templateName}')`
  }
}

function createQueries(components: Record<string, Type<any>>, angularVersion: number) {
  const componentNames = Object.keys(components)

  return componentNames.map((name) => {
    const selector = getComponentAnnotations(components[name]).selector
    const templateName = kebabToCamelCase(selector)
    return createViewChildString(name, templateName, angularVersion)
  })
}


export function createComponentString(components: Record<string, Type<any>>, angularVersion: number) {
  const queries = createQueries(components, angularVersion)

  const componentString = `
    @Component({
      selector: '${remoteViewSelector}',
      template,
      inputs: ['loadParams'],
      queries: {
        ${queries.join(',\n        ')},
      },
    })
    export class BeagleRemoteView extends AbstractBeagleRemoteView {
      constructor(
        beagleProvider: BeagleProvider,
        ngZone: NgZone,
        changeDetector: ChangeDetectorRef,
      ) {
        // @ts-ignore
        super(beagleProvider, ngZone, changeDetector)
      }
    }
  `

  return removeExtraIndentation(componentString, 4)
}
