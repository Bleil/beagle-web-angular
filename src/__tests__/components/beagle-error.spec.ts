/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { TestBed, async, ComponentFixture } from '@angular/core/testing'
import { BeagleErrorComponent } from '../../components/beagle-error/beagle-error.component'
import { BeagleTextComponent } from '../../components/beagle-text/beagle-text.component'

let component: BeagleErrorComponent
let fixture: ComponentFixture<BeagleErrorComponent>
let compiled: HTMLElement

describe('BeagleErrorComponent - empty', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        BeagleTextComponent,
        BeagleErrorComponent,
      ],
    }).compileComponents()
  }))

  beforeEach(() => {
    fixture = TestBed.createComponent(BeagleErrorComponent)
    compiled = fixture.debugElement.nativeElement

    component = fixture.componentInstance
    component.retry = jest.fn()
    component.errors = []
    component.showingDetails = false

    fixture.detectChanges()
  })

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot()
  })

  it('Should not render details when no error is provided', () => {
    const showDetailsButton = compiled.querySelectorAll('.show-details-button')
    const errorsDetailedContainer = compiled.querySelectorAll('.errors-detailed-container')

    expect(showDetailsButton.length).toBe(0)
    expect(errorsDetailedContainer.length).toBe(0)
  })

  it('Should call retry function on click', () => {
    const button = compiled.querySelector('.retry-button') as HTMLButtonElement
    button.click()

    expect(component.retry).toHaveBeenCalled()
  })
})

describe('BeagleErrorComponent - with errors', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        BeagleTextComponent,
        BeagleErrorComponent,
      ],
    }).compileComponents()
  }))

  beforeEach(() => {
    fixture = TestBed.createComponent(BeagleErrorComponent)
    compiled = fixture.debugElement.nativeElement

    component = fixture.componentInstance
    component.retry = jest.fn()
    component.errors = [{ message: 'First Message' }, { message: 'Second Message' }]
    component.showingDetails = false

    fixture.detectChanges()
  })

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot()
  })

  it('Should call retry function on click', () => {
    const button = compiled.querySelector('.retry-button') as HTMLButtonElement
    button.click()

    expect(component.retry).toHaveBeenCalled()
  })

  it('Should render details when error list is provided', () => {
    const showDetailsButton = compiled.querySelectorAll('.show-details-button')
    const errorsDetailedContainer = compiled.querySelectorAll('.errors-detailed-container')

    expect(showDetailsButton.length).toBe(1)
    expect(errorsDetailedContainer.length).toBe(1)

    const errorMessages = errorsDetailedContainer[0].querySelectorAll('p')

    expect(errorMessages.length).toBe(2)
  })

  it('Should not add the class when the show more button was not clicked', () => {
    const errorsDetailedContainer = compiled.querySelector('.errors-detailed-container')
    const showContainer = errorsDetailedContainer?.querySelectorAll('.show')

    expect(showContainer?.length).toBe(0)
  })

  it('Should add the class when the show more button was clicked', () => {
    const button = compiled.querySelector('.show-details-button') as HTMLButtonElement
    button.click()
    fixture.detectChanges()

    const errorsDetailedContainer = compiled.querySelector('.errors-detailed-container')

    expect(component.showingDetails).toBeTruthy()
    expect(errorsDetailedContainer?.classList.contains('show')).toBeTruthy()

    button.click()
    fixture.detectChanges()

    expect(component.showingDetails).toBeFalsy()
    expect(errorsDetailedContainer?.classList.contains('show')).toBeFalsy()
  })
})
