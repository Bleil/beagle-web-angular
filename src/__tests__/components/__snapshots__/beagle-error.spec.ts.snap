// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeagleErrorComponent - empty should match snapshot 1`] = `
<beagle-error
  errors={[Function Array]}
  retry={[Function Function]}
  showingDetails="false"
>
  <beagle-text
    text="Sorry!"
    textcolor="red"
  >
    <p
      class="beagle-text"
      style="color: red; text-align: inherit;"
    >
      Sorry!
    </p>
  </beagle-text><beagle-text
    text="An unexpected error happened while loading your page."
  >
    <p
      class="beagle-text"
      style="text-align: inherit;"
    >
      An unexpected error happened while loading your page.
    </p>
  </beagle-text><button
    class="retry-button"
    type="button"
  >
    Retry
  </button>
</beagle-error>
`;

exports[`BeagleErrorComponent - with errors should match snapshot 1`] = `
<beagle-error
  errors={[Function Array]}
  retry={[Function Function]}
  showingDetails="false"
>
  <beagle-text
    text="Sorry!"
    textcolor="red"
  >
    <p
      class="beagle-text"
      style="color: red; text-align: inherit;"
    >
      Sorry!
    </p>
  </beagle-text><beagle-text
    text="An unexpected error happened while loading your page."
  >
    <p
      class="beagle-text"
      style="text-align: inherit;"
    >
      An unexpected error happened while loading your page.
    </p>
  </beagle-text><button
    class="retry-button"
    type="button"
  >
    Retry
  </button><button
    class="show-details-button"
    type="button"
  >
     Show details ▼ 
  </button><section
    class="errors-detailed-container"
  >
    
    <p
      class="error-message"
    >
      <span
        class="error-indicator"
      >
        ►
      </span>
       ERROR: First Message 
    </p>
    <p
      class="error-message"
    >
      <span
        class="error-indicator"
      >
        ►
      </span>
       ERROR: Second Message 
    </p>
  </section>
</beagle-error>
`;
